core/blockchain.go

```diff
@@ -2381,7 +2358,21 @@ func (bc *BlockChain) UpdateValidatorSnapshots(
 
 	allValidators = append(allValidators, newValidators...)
 
-	return bc.writeValidatorSnapshots(batch, allValidators, epoch, state)
+	// Read all validator's current data and snapshot them
+	for i := range allValidators {
```

Comment:
We can do this concurrently

Reply:
Leveldb batch is not thread safe https://github.com/google/leveldb/blob/master/doc/index.md#concurrency

Reply:
interesting - how much benefit is there to the batch though vs concurrent write?

Reply:
Batch is not for mainly for speed, but for atomicity